src/doc/octree_map.cpp

```diff
@@ -0,0 +1,248 @@
+// Aseprite
+// Copyright (c) 2019 Igara Studio S.A.
+//
+// This file is released under the terms of the MIT license.
+// Read LICENSE.txt for more information.
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "doc/octree_map.h"
+
+#include "doc/palette.h"
+
+#define MID_VALUE_COLOR ((rgba_r_mask / 2) + 1)
+
+namespace doc {
+
+void OctreeNode::addColor(color_t c, int level, OctreeNode* parent,
+                          int paletteIndex, int levelDeep)
+{
+  m_parent = parent;
+  if (level >= levelDeep) {
+    m_leafColor.add(c);
+    m_paletteIndex = paletteIndex;
+    return;
+  }
+  int index = getOctet(c, level);
+  if (!m_children) {
+    m_children.reset(new std::array<OctreeNode, 8>());
+  }
+  (*m_children)[index].addColor(c, level + 1, this, paletteIndex, levelDeep);
+}
+
+void OctreeNode::fillOrphansNodes(const Palette* palette,
+                                  color_t upstreamBranchColor, int level)
+{
+  for (int i=0; i<8; i++) {
+    if ((*m_children)[i].m_children)
+      (*m_children)[i].fillOrphansNodes(
+        palette,
+        upstreamBranchColor + octetToBranchColor(i, level),
+        level + 1);
+    else if (!((*m_children)[i].isLeaf())) {
+      // Here the node IS NOT a Leaf and HAS NOT children
+      // So, we must assign palette index to the current node
+      // to fill the "map holes" (i.e "death tree branches")
+      int currentBranchColorAdd = octetToBranchColor(i, level);
+      int midColorAdd = MID_VALUE_COLOR >> (level + 1);
+      midColorAdd = (midColorAdd << rgba_r_shift)
+                  + (midColorAdd << rgba_g_shift)
+                  + (midColorAdd << rgba_b_shift);
+      color_t branchColorMed = rgba_a_mask
+                             + upstreamBranchColor
+                             + currentBranchColorAdd
+                             + midColorAdd;
```

Comment:
I think all these `+` operations should be `|` (in practice the result should be the same, but prefer `|` if you are dealing with bits instead of integer values).